imap/caldav_alarm.c

```diff
@@ -865,20 +838,169 @@ static void process_one_record(struct mailbox *mailbox, uint32_t imap_uid,
+    int r = 0;
+
+    syslog(LOG_DEBUG, "processing future delivery");
+
+    if ((record->internal_flags & FLAG_INTERNAL_EXPUNGED) ||
+        (record->system_flags & FLAG_DELETED))
+        goto done;
+
+    /* Load message */
+    message_t *m = message_new_from_record(mailbox, record);
+    const char *fname;
+    r = message_get_fname(m, &fname);
+    if (r) {
+        syslog(LOG_ERR, "process_futurerelease: can't get fname for %s:%u",
+               mailbox->name, record->uid);
+        goto done;
+    }
+
+    int msgfd = open(fname, 0);
+    if (msgfd == -1) {
+        syslog(LOG_ERR, "process_futurerelease: can't open %s: %m", fname);
+        goto done;
+    }
+
+    struct stat sbuf;
+    if (fstat(msgfd, &sbuf) == -1) {
+        syslog(LOG_ERR, "process_futurerelease: can't fstat %s: %m", fname);
+        close(msgfd);
+        goto done;
+    }
+
+    buf_init_mmap(&buf, 1, msgfd, fname, sbuf.st_size, mailbox->name);
+    close(msgfd);
+    if (!buf_len(&buf)) {
+        syslog(LOG_ERR, "process_futurerelease: can't mmap %s: %m", fname);
+        goto done;
+    }
+
+    /* Parse the submission object from the first header field */
+    json_error_t jerr;
+    size_t hdrlen = strlen(JMAP_SUBMISSION_HDR) + 1; /* +1 for ':' */
+    submission = json_loadb(buf_base(&buf) + hdrlen, buf_len(&buf) - hdrlen,
+                            JSON_DISABLE_EOF_CHECK, &jerr);
+    if (!submission) {
+        syslog(LOG_ERR,
+               "process_futurerelease: failed to parse submission obj");
+        goto done;
+    }
+    envelope = json_object_get(submission, "envelope");
+    identity = json_object_get(submission, "identityId");
+
+    /* Trim submission header field */
+    const char *p = strstr(buf_base(&buf), "}\r\n") + 3; /* +3 for '}\r\n' */
+    buf_remove(&buf, 0, p - buf_base(&buf));
+
+    /* Open the SMTP connection */
+    smtpclient_t *sm = NULL;
+    r = smtpclient_open(&sm);
+    if (r) {
+        syslog(LOG_ERR, "smtpclient_open failed: %s", error_message(r));
+        goto done;
+    }
+    smtpclient_set_auth(sm, json_string_value(identity));
+
+    /* Prepare envelope */
+    smtp_envelope_t smtpenv = SMTP_ENVELOPE_INITIALIZER;
+    jmap_emailsubmission_envelope_to_smtp(&smtpenv, envelope);
+
+    /* Send message */
+    r = smtpclient_send(sm, &smtpenv, &buf);
+    if (r) {
+        syslog(LOG_ERR, "smtpclient_send failed: %s", error_message(r));
+        goto done;
+    }
+    smtp_envelope_fini(&smtpenv);
+    smtpclient_close(&sm);
+
+    /* Expunge the resource */
+    record->internal_flags |= FLAG_INTERNAL_EXPUNGED;
+    r = mailbox_rewrite_index_record(mailbox, record);
+    if (r) {
+        syslog(LOG_ERR, "expunging record (%s:%u) failed: %s",
+               mailbox->name, record->uid, error_message(r));
+    }
+
+  done:
+    if (r) {
+        /* XXX  Error handling: retry, DSN, both, neither? */
+        caldav_alarm_delete_record(mailbox->name, record->uid);
```

Comment:
Yeah, this is a bit annoying.  A dav_reconstruct or similar would presumably re-add this record, because we haven't updated the record at all.  We do want to think about this some more.